pkg/sources/nats/nats.go

```diff
@@ -0,0 +1,265 @@
+	"github.com/numaproj/numaflow/pkg/isb/forward"
+	metricspkg "github.com/numaproj/numaflow/pkg/metrics"
+	"github.com/numaproj/numaflow/pkg/shared/logging"
+	sharedutil "github.com/numaproj/numaflow/pkg/shared/util"
+	"github.com/numaproj/numaflow/pkg/udf/applier"
+	"github.com/numaproj/numaflow/pkg/watermark/fetch"
+	"github.com/numaproj/numaflow/pkg/watermark/processor"
+	"github.com/numaproj/numaflow/pkg/watermark/publish"
+	"github.com/numaproj/numaflow/pkg/watermark/store"
+)
+
+type natsSource struct {
+	name         string
+	pipelineName string
+	logger       *zap.SugaredLogger
+
+	natsConn *natslib.Conn
+	sub      *natslib.Subscription
+
+	bufferSize  int
+	messages    chan *isb.ReadMessage
+	readTimeout time.Duration
+
+	cancelfn  context.CancelFunc
+	forwarder *forward.InterStepDataForward
+	// source watermark publisher
+	sourcePublishWM publish.Publisher
+}
+
+func New(vertexInstance *dfv1.VertexInstance, writers []isb.BufferWriter, fetchWM fetch.Fetcher, publishWM map[string]publish.Publisher, publishWMStores store.WatermarkStorer, opts ...Option) (*natsSource, error) {
+	n := &natsSource{
+		name:         vertexInstance.Vertex.Spec.Name,
+		pipelineName: vertexInstance.Vertex.Spec.PipelineName,
+		bufferSize:   1000,            // default size
+		readTimeout:  1 * time.Second, // default timeout
+	}
+	for _, o := range opts {
+		operr := o(n)
+		if operr != nil {
+			return nil, operr
+		}
+	}
+	if n.logger == nil {
+		n.logger = logging.NewLogger()
+	}
+	n.messages = make(chan *isb.ReadMessage, n.bufferSize)
+
+	destinations := make(map[string]isb.BufferWriter, len(writers))
+	for _, w := range writers {
+		destinations[w.GetName()] = w
+	}
+	forwardOpts := []forward.Option{forward.WithVertexType(dfv1.VertexTypeSource), forward.WithLogger(n.logger)}
+	if x := vertexInstance.Vertex.Spec.Limits; x != nil {
+		if x.ReadBatchSize != nil {
+			forwardOpts = append(forwardOpts, forward.WithReadBatchSize(int64(*x.ReadBatchSize)))
+		}
+	}
+	forwarder, err := forward.NewInterStepDataForward(vertexInstance.Vertex, n, destinations, forward.All, applier.Terminal, fetchWM, publishWM, forwardOpts...)
+	if err != nil {
+		n.logger.Errorw("Error instantiating the forwarder", zap.Error(err))
+		return nil, err
+	}
+	n.forwarder = forwarder
+	ctx, cancel := context.WithCancel(context.Background())
+	n.cancelfn = cancel
+	entityName := fmt.Sprintf("%s-%d", vertexInstance.Vertex.Name, vertexInstance.Replica)
+	processorEntity := processor.NewProcessorEntity(entityName)
+	n.sourcePublishWM = publish.NewPublish(ctx, processorEntity, publishWMStores, publish.IsSource(), publish.WithDelay(vertexInstance.Vertex.Spec.Watermark.GetMaxDelay()))
+
+	source := vertexInstance.Vertex.Spec.Source.Nats
+	opt := []natslib.Option{
+		natslib.MaxReconnects(-1),
+		natslib.ReconnectWait(3 * time.Second),
+		natslib.DisconnectHandler(func(c *natslib.Conn) {
+			n.logger.Info("Nats disconnected")
+		}),
+		natslib.ReconnectHandler(func(c *natslib.Conn) {
```

Comment:
we are using the default reconnector from nats.go, right?

Reply:
Yes. Our auto reconnection only works for JetStream.